File: 2024/matlab/aoc2.m
input_path = fullfile(get_input_root,'day_2.txt');

x = read_txt(input_path);
x = strip(x);
lines = strsplit(x,crlf);

%part1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
s = zeros(numel(lines),1);
for i1 = 1:numel(lines)
   v = lines{i1};
   ns = strsplit(v,' ');
   n = str2double(ns);
   dn = diff(n);
   if all(sign(dn) >0)||all(sign(dn) <0)
       if min(abs(dn))>=1&&max(abs(dn))<=3
           s(i1) = 1;
       end
   end
end
sum(s)

%part2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
s = zeros(numel(lines),1);
for i1 = 1:numel(lines)
    v = lines{i1};
    ns = strsplit(v,' ');
    n = str2double(ns);
    for i2 = 1:numel(n)
        ni2 = n;
        ni2(i2) = [];
        dn = diff(ni2);
        if all(sign(dn) >0)||all(sign(dn) <0)
            if min(abs(dn))>=1&&max(abs(dn))<=3
                s(i1) = 1;
            end
        end
    end
end
sum(s)

File: 2024/python/aoc2.py
# -*- coding: utf-8 -*-
import numpy as np

input_path = '..\inputs\day_2.txt'
txt = open(input_path).read()

lines = txt.split('\n')

r = [[int(x) for x in l.split()] for l in lines]

def is_safe(n):
    nd = np.diff(n)
    return (np.all(nd < 0) or np.all(nd > 0)) and (np.all(abs(nd)>=1) and np.all(abs(nd)<=3))

#part1 ########################################################################
print(sum([is_safe(n) for n in r]))

#part2 ########################################################################
print(sum(any(is_safe(np.delete(n,j)) for j in range(len(n))) for n in r))